### ISSUE TITLE
Recorded audio can't be read as an .mp3 file

### ISSUE THREAD

**Issue by B4PT0R** [theevann/streamlit-audiorecorder#8]
Hi! (again)

In your example you suggest writing audio.tobytes() to an mp3 file.

I tested this, the mp3 file doesn't play in vlc. But if written in a .wav file, the wav file reads without problem.

This suggests that the bytes data outputted by your audiorecorder is intrinsically encoded in wav format.

Yet, when trying to convert the wav audio file (or bytes) to an AudioSegment, pydub crashes, warning me that the file isn't in proper wav format (even though it reads fine in vlc!). This is probably because the file doesn't have the proper header a wav file should have.

So I checked the source code of your react component. I'm not a front-end specialist, but when you process the data output of the MediaRecorder object, you write : 
```typescript
 recorder.ondataavailable = async ({data}) => {
        const audioData = new Blob([data], {type: "audio/mp3"});
        const audioData_str = (await BlobToDataURL(audioData)).replace(/^data:.+?base64,/, "");
        Streamlit.setComponentValue(audioData_str);
      }
```
When creating the blob, you force its type to "audio/mp3". This will result in setting the header of the blob as the one of an mp3 file.
But the data itself coming out of the recorder is likely not in mp3 format (I guess it is in wav format), and just forcing the type of the blob won't achieve any conversion. You just put wav bytes data into a blob with mp3 header. This results in crashes when having to convert the content with other audio tools, such as pydub, the wave module, speech_recognition etc...

Some audio tools such as vlc or st.audio will sometimes ignore the header to process only the bytes content. That's why reading the data still works in some cases : st.audio thinks you input wav bytes into it so it plays it as a wav. But it will make more advanced audio tools depending on the header crash.

To solve this issue, you probably just have to change "audio/mp3" to "audio/wav" when creating the blob.

This should do the trick !

**Comment by theevann** [comment]
Hey !
Thank you for your detailed message.

I did some digging 2 weeks ago for a whole day and came to similar conclusions.
I think the returned data are the bytes encoding of data + some header.

The default audio format seems to be webm and not wav but I still have to confirm that, if so a simple python-side conversion should do the trick. 

Doing `new Blob([data], {type: "audio/mp3"});` does not actually convert `data` to mp3 as you said, just add some headers (which in hindsight is quite logical). You can specify the desired format in the MediaRecorder constructor but when I tried wav and mp3, both failed. I came to the same conclusion regarding vlc being able to read it.

My next steps as of now whenever I have half a day for this is:
- to push the upgrade to the new version streamlit,
- to confirm this is webm format or find out what it is and do some audio conversion python-side to return some pydub,
- to allow passing the sampling frequency and more if possible

If I can't figure out the bytes thing, I am considering looking into a lower level interface of the browser to have more flexibility.

Sorry for the delay, thanks again.

**Comment by theevann** [comment]
The [new version](https://github.com/theevann/streamlit-audiorecorder/releases/tag/v0.0.3.post1) should solve this.

**Comment by B4PT0R** [comment]
Hi,

Thanks for the reply.

Waiting for your response, I've spent the week working on my own version based on your initial work and figured out how to make the appropriate audio conversions on the front end's side, depending on the audio format returned by the browser (generaly audio/webm yes, but it may vary from one browser to another). The widget now converts the audio data received by the MediaRecorder directly on the front end's side, and reliably returns audio/wav mono bytes along with the sample rate and sample width of the audio which simplifies further audio processing in case the user needs these informations, while avoiding the pydub dependency.

I've made it so that passing the audio to the speech_recognition module is straightforward, you just create an sr.AudioData object passing it the bytes, sample_rate and sample_width, and run the recognizer on this object, without more overhead. This allowed to create a dedicated function that performs STT directly on the audio recorded (which adds a speech_recognition dependency, but I think this is how most users will use the widget anyway).

I also struggled to get the theming right to look just like the native st.button style (I noticed that st.button border and background color are not colors passed int the theme prop, but rather some kind of weighted average of these colors. I iterated a bit with the TinyColor package to get as close as possible, and the result is now satisfying enough.

Added a "use_container_width" parameter for the button, like for st.button.

Also added a "just_once" parameter for convenience, which allow to control if the audio/text will be returned on every rerun of the app once it has been recorded, or only once and then None (to avoid triggering a subsequent logic several times on the same audio / text, which can be tricky to avoid otherwise).

Fell free to have a look at it [here](https://github.com/B4PT0R/streamlit-mic-recorder).

It's not on PyPI yet. Never created a PyPI package before so I need to figure out how to do it properly.

Since our versions are mostly duplicates, why not merge them as one ?

Feel free to update your repo with the last modifications I made (in case you think they're relevant), and publish the updated package.

Have a good day!

Cheers
